<a href="https://colab.research.google.com/github/ambreenraheem/PGD-Machine-Learning-2024-25/blob/main/PK_import_MLproject.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Ambreen Abdul raheem**
#### The Data analyst and Web Developer

#### Project: **Pakistan Import Machine Learning Project**
**[GitHub](https://github.com/ambreenraheem)**

**[LinkedIn](https://www.linkedin.com/in/ambreen-abdul-raheem/)**

**[YouTube](https://www.youtube.com/@AmbreenAbdulRaheem-y8m)**

**[Discort](?)** will add link later

**Step # 01: Importing libraries**

In [123]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

**Step # 02: Loading Dataset of Pakistan Import**

In [124]:
df_import=pd.read_excel("Master_Import_Merged_2021_2025.xlsx")
df_import.head(20)

,#,CATEGORY,CLASS TYPE,PRODUCT,UNIT,QUANTITY,RUPEES,DOLLARS,% QUANTITY,% RUPEES,% DOLLARS,STATUS,MONTH,YEAR
0,3,FOOD GROUP,NaN,"MILK,CREAM & MILK FOOD FOR INFANTS",MT,4199.0,1989,12993,-36.16,-39.25,-38.08,Import,April,2021
1,4,FOOD GROUP,NaN,WHEAT UNMILLED,MT,0.0,0,0,-100.00,-100.00,-100.00,Import,April,2021
2,5,FOOD GROUP,NaN,DRY FRUITS & NUTS,MT,7191.0,630,4116,-19.96,-17.65,-16.09,Import,April,2021
3,6,FOOD GROUP,NaN,TEA,MT,22779.0,8095,52882,-3.04,-7.00,-5.20,Import,April,2021
4,7,FOOD GROUP,NaN,SPICES,MT,14881.0,3736,24406,-13.02,24.49,26.88,Import,April,2021
5,8,FOOD GROUP,NaN,SOYABEAN OIL,MT,13300.0,2154,14071,100.00,100.00,100.00,Import,April,2021
6,9,FOOD GROUP,NaN,PALM OIL,MT,275371.0,43055,281264,-0.25,0.14,2.07,Import,April,2021
7,10,FOOD GROUP,NaN,SUGAR,MT,773.0,71,464,-6.42,-4.05,-2.73,Import,April,2021
8,11,FOOD GROUP,NaN,PULSES (LEGUMINOUS VEGETABLES),MT,155747.0,14624,95534,26.62,32.69,35.26,Import,April,2021
9,12,FOOD GROUP,NaN,ALL OTHERS FOOD ITEMS,NaN,NaN,44735,292239,NaN,9.61,11.73,Import,April,2021


**Step # 03: Some important info of the dataset**
- Data's length
- Data's info
- Data's Null Values

In [125]:
df_import["CATEGORY"].count()

np.int64(2805)

In [126]:
df_import.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2805 entries, 0 to 2804
Data columns (total 14 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   #           2805 non-null   int64  
 1   CATEGORY    2805 non-null   object 
 2   CLASS TYPE  510 non-null    object 
 3   PRODUCT     2805 non-null   object 
 4   UNIT        1326 non-null   object 
 5   QUANTITY    1328 non-null   float64
 6   RUPEES      2805 non-null   int64  
 7   DOLLARS     2805 non-null   int64  
 8   % QUANTITY  1326 non-null   float64
 9   % RUPEES    2805 non-null   float64
 10  % DOLLARS   2805 non-null   float64
 11  STATUS      2805 non-null   object 
 12  MONTH       2805 non-null   object 
 13  YEAR        2805 non-null   int64  
dtypes: float64(4), int64(4), object(6)
memory usage: 306.9+ KB


In [127]:
df_import.isnull().sum()

,0
#,0
CATEGORY,0
CLASS TYPE,2295
PRODUCT,0
UNIT,1479
QUANTITY,1477
RUPEES,0
DOLLARS,0
% QUANTITY,1479
% RUPEES,0


**Step # 04: Let's handle null values and prepare the data for the ML model.**

- first [CLASS TYPE] feature
1. Categorical data and rename the column name, it has a space
2. Total missing values = 2,295
3. We cannot remove all null values because it may cause issues in the dataset
4. Replace all null values with the keyword "unknown"

In [128]:
df_import1 = df_import.rename(columns={"CLASS TYPE": "CLASS_TYPE"})

In [129]:
df_import1

,#,CATEGORY,CLASS_TYPE,PRODUCT,UNIT,QUANTITY,RUPEES,DOLLARS,% QUANTITY,% RUPEES,% DOLLARS,STATUS,MONTH,YEAR
0,3,FOOD GROUP,NaN,"MILK,CREAM & MILK FOOD FOR INFANTS",MT,4199.0,1989,12993,-36.16,-39.25,-38.08,Import,April,2021
1,4,FOOD GROUP,NaN,WHEAT UNMILLED,MT,0.0,0,0,-100.00,-100.00,-100.00,Import,April,2021
2,5,FOOD GROUP,NaN,DRY FRUITS & NUTS,MT,7191.0,630,4116,-19.96,-17.65,-16.09,Import,April,2021
3,6,FOOD GROUP,NaN,TEA,MT,22779.0,8095,52882,-3.04,-7.00,-5.20,Import,April,2021
4,7,FOOD GROUP,NaN,SPICES,MT,14881.0,3736,24406,-13.02,24.49,26.88,Import,April,2021
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2800,3464,MISCELLANEOUS GROUP,NaN,RUBBER TYRES & TUBES,NO,504085.0,3177,11343,-5.44,37.65,37.27,Import,March,2025
2801,3465,MISCELLANEOUS GROUP,NaN,WOOD & CORK,NaN,NaN,2456,8768,NaN,-6.05,-6.30,Import,March,2025
2802,3466,MISCELLANEOUS GROUP,NaN,JUTE,MT,209.0,50,177,-89.27,-88.37,-88.42,Import,March,2025
2803,3467,MISCELLANEOUS GROUP,NaN,PAPER & PAPER BOARD & MANUF THEREOF,MT,35597.0,10051,35885,9.14,12.43,12.11,Import,March,2025


In [130]:
df_import1['CLASS_TYPE']=df_import1['CLASS_TYPE'].fillna('unknown')

In [131]:
df_import1['CLASS_TYPE'].isnull().sum()

np.int64(0)

In [132]:
df_import1.head(20)

,#,CATEGORY,CLASS_TYPE,PRODUCT,UNIT,QUANTITY,RUPEES,DOLLARS,% QUANTITY,% RUPEES,% DOLLARS,STATUS,MONTH,YEAR
0,3,FOOD GROUP,unknown,"MILK,CREAM & MILK FOOD FOR INFANTS",MT,4199.0,1989,12993,-36.16,-39.25,-38.08,Import,April,2021
1,4,FOOD GROUP,unknown,WHEAT UNMILLED,MT,0.0,0,0,-100.00,-100.00,-100.00,Import,April,2021
2,5,FOOD GROUP,unknown,DRY FRUITS & NUTS,MT,7191.0,630,4116,-19.96,-17.65,-16.09,Import,April,2021
3,6,FOOD GROUP,unknown,TEA,MT,22779.0,8095,52882,-3.04,-7.00,-5.20,Import,April,2021
4,7,FOOD GROUP,unknown,SPICES,MT,14881.0,3736,24406,-13.02,24.49,26.88,Import,April,2021
5,8,FOOD GROUP,unknown,SOYABEAN OIL,MT,13300.0,2154,14071,100.00,100.00,100.00,Import,April,2021
6,9,FOOD GROUP,unknown,PALM OIL,MT,275371.0,43055,281264,-0.25,0.14,2.07,Import,April,2021
7,10,FOOD GROUP,unknown,SUGAR,MT,773.0,71,464,-6.42,-4.05,-2.73,Import,April,2021
8,11,FOOD GROUP,unknown,PULSES (LEGUMINOUS VEGETABLES),MT,155747.0,14624,95534,26.62,32.69,35.26,Import,April,2021
9,12,FOOD GROUP,unknown,ALL OTHERS FOOD ITEMS,NaN,NaN,44735,292239,NaN,9.61,11.73,Import,April,2021


- second, we have the [QUANTITY] and [%QUANTITY] features:
1. Both features contain missing values and are important for the dataset.
2. Total missing values: [QUANTITY] = 1,477 and [%QUANTITY] = 1,479
3. We need to replace the missing values using groupby on the [PRODUCT] feature.
4. The [PRODUCT] feature is required to estimate/derive quantity values.

In [133]:
# QUANTITY
df_import1['QUANTITY'] = df_import1.groupby(['PRODUCT', 'YEAR'])['QUANTITY']\
                                 .transform(lambda x: x.fillna(x.median()))

In [134]:
df_import1['QUANTITY'] = df_import1['QUANTITY'].fillna(df_import1['QUANTITY'].median())

# % QUANTITY
df_import1['% QUANTITY'] = df_import1['% QUANTITY'].fillna(0)

In [135]:
df_import1.head(20)

,#,CATEGORY,CLASS_TYPE,PRODUCT,UNIT,QUANTITY,RUPEES,DOLLARS,% QUANTITY,% RUPEES,% DOLLARS,STATUS,MONTH,YEAR
0,3,FOOD GROUP,unknown,"MILK,CREAM & MILK FOOD FOR INFANTS",MT,4199.0,1989,12993,-36.16,-39.25,-38.08,Import,April,2021
1,4,FOOD GROUP,unknown,WHEAT UNMILLED,MT,0.0,0,0,-100.00,-100.00,-100.00,Import,April,2021
2,5,FOOD GROUP,unknown,DRY FRUITS & NUTS,MT,7191.0,630,4116,-19.96,-17.65,-16.09,Import,April,2021
3,6,FOOD GROUP,unknown,TEA,MT,22779.0,8095,52882,-3.04,-7.00,-5.20,Import,April,2021
4,7,FOOD GROUP,unknown,SPICES,MT,14881.0,3736,24406,-13.02,24.49,26.88,Import,April,2021
5,8,FOOD GROUP,unknown,SOYABEAN OIL,MT,13300.0,2154,14071,100.00,100.00,100.00,Import,April,2021
6,9,FOOD GROUP,unknown,PALM OIL,MT,275371.0,43055,281264,-0.25,0.14,2.07,Import,April,2021
7,10,FOOD GROUP,unknown,SUGAR,MT,773.0,71,464,-6.42,-4.05,-2.73,Import,April,2021
8,11,FOOD GROUP,unknown,PULSES (LEGUMINOUS VEGETABLES),MT,155747.0,14624,95534,26.62,32.69,35.26,Import,April,2021
9,12,FOOD GROUP,unknown,ALL OTHERS FOOD ITEMS,NaN,32568.0,44735,292239,0.00,9.61,11.73,Import,April,2021


- third [UNIT] feature
1. Categorical data
2. Total missimg Values = 1,479
3. It has three values (MT, KG, NO), here we have to handle data
4. Prepare the dataset for the ML Model
5. Best for ML (separating the data into [df_weight= for ML, df_count= separate analysis])

In [136]:
df_import1['UNIT_TYPE'] = df_import1['UNIT'].map({
    'MT': 'WEIGHT',
    'KG': 'WEIGHT',
    'NO': 'COUNT'
})

### Imputing Missing `UNIT_TYPE` based on `PRODUCT`
("Imputed missing 'UNIT_TYPE' values based on product mode, and then global mode if necessary.")

In [149]:
# Impute UNIT_TYPE based on the mode within each PRODUCT group
df_import1['UNIT_TYPE'] = df_import1.groupby('PRODUCT')['UNIT_TYPE'].transform(lambda x: x.fillna(x.mode()[0]) if not x.mode().empty else x)

# Handle any remaining NaN values (e.g., if an entire product group had no UNIT_TYPE defined)
# We'll fill these with the overall mode of UNIT_TYPE
if df_import1['UNIT_TYPE'].isnull().any():
    global_unit_type_mode = df_import1['UNIT_TYPE'].mode()[0]
    df_import1['UNIT_TYPE'].fillna(global_unit_type_mode, inplace=True)

Now, let's verify if there are any remaining null values in the `UNIT_TYPE` column after this more sophisticated imputation.

Let's also check the overall null values for `df_import1` to ensure consistency after these adjustments.

In [151]:
df_import1.isnull().sum()

,0
#,0
CATEGORY,0
CLASS_TYPE,0
PRODUCT,0
UNIT,1479
QUANTITY,0
RUPEES,0
DOLLARS,0
% QUANTITY,0
% RUPEES,0


In [152]:
df_import1.head(20)

,#,CATEGORY,CLASS_TYPE,PRODUCT,UNIT,QUANTITY,RUPEES,DOLLARS,% QUANTITY,% RUPEES,% DOLLARS,STATUS,MONTH,YEAR,UNIT_TYPE
0,3,FOOD GROUP,unknown,"MILK,CREAM & MILK FOOD FOR INFANTS",MT,4199.0,1989,12993,-36.16,-39.25,-38.08,Import,April,2021,WEIGHT
1,4,FOOD GROUP,unknown,WHEAT UNMILLED,MT,0.0,0,0,-100.00,-100.00,-100.00,Import,April,2021,WEIGHT
2,5,FOOD GROUP,unknown,DRY FRUITS & NUTS,MT,7191.0,630,4116,-19.96,-17.65,-16.09,Import,April,2021,WEIGHT
3,6,FOOD GROUP,unknown,TEA,MT,22779.0,8095,52882,-3.04,-7.00,-5.20,Import,April,2021,WEIGHT
4,7,FOOD GROUP,unknown,SPICES,MT,14881.0,3736,24406,-13.02,24.49,26.88,Import,April,2021,WEIGHT
5,8,FOOD GROUP,unknown,SOYABEAN OIL,MT,13300.0,2154,14071,100.00,100.00,100.00,Import,April,2021,WEIGHT
6,9,FOOD GROUP,unknown,PALM OIL,MT,275371.0,43055,281264,-0.25,0.14,2.07,Import,April,2021,WEIGHT
7,10,FOOD GROUP,unknown,SUGAR,MT,773.0,71,464,-6.42,-4.05,-2.73,Import,April,2021,WEIGHT
8,11,FOOD GROUP,unknown,PULSES (LEGUMINOUS VEGETABLES),MT,155747.0,14624,95534,26.62,32.69,35.26,Import,April,2021,WEIGHT
9,12,FOOD GROUP,unknown,ALL OTHERS FOOD ITEMS,NaN,32568.0,44735,292239,0.00,9.61,11.73,Import,April,2021,WEIGHT


In [153]:
df_weight = df_import1[df_import1['UNIT_TYPE'] == 'WEIGHT']
df_count  = df_import1[df_import1['UNIT_TYPE'] == 'COUNT']

In [154]:
df_weight['UNIT_TYPE'].value_counts()

,count
UNIT_TYPE,
WEIGHT,2754


In [155]:
df_count['UNIT_TYPE'].value_counts()

,count
UNIT_TYPE,
COUNT,51


In [156]:
df_import1['UNIT_TYPE'].value_counts()

,count
UNIT_TYPE,
WEIGHT,2754
COUNT,51


In [157]:
df_import1['UNIT_TYPE'].isnull().sum()

np.int64(0)